### PR TITLE
Fix zkVM IPC deadlock

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR fixes the halt caused by deadlocks between the zkVM and the preimage oracle, by setting the `DIRECT` flag in the pipes that we pass to the oracle process. There isn't a nice API for this, so there's some low-level detail here, but the zkVM now runs end-to-end unsupervised!

Massive thanks to @jspada talking over the problem and proposing the solution :heart:

### CODE REVIEW

File: optimism/src/preimage_oracle.rs [hunk 1]
```diff
@@ -24,6 +24,57 @@ pub struct ReadWrite<R, W> {
...
+}
+
+#[cfg(any(target_os = "ios", target_os = "macos", target_os = "haiku"))]
+pub fn create_pipe() -> std::io::Result<(PipeReader, PipeWriter)> {
+    let mut fds: [libc::c_int; 2] = [0; 2];
+    let res = unsafe { libc::pipe(fds.as_mut_ptr()) };
+    if res != 0 {
+        return Err(std::io::Error::last_os_error());
+    }
+    // It appears that Mac and friends don't have DIRECT. Oh well. Don't use a Mac.
```
Comment (dannywillems) [inline optimism/src/preimage_oracle.rs:55]: +1

File: optimism/src/preimage_oracle.rs [hunk 2]
```diff
@@ -24,6 +24,57 @@ pub struct ReadWrite<R, W> {
...
+            PipeWriter::from_raw_fd(fds[1]),
+        ))
+    }
+}
+
+#[cfg(windows)]
+pub fn create_pipe() -> std::io::Result<(PipeReader, PipeWriter)> {
+    os_pipe::pipe()
+}
+
```
Comment (rbonichon) [inline optimism/src/preimage_oracle.rs:77]: Do we actually hope to run any of this on Windows? If not this would remove the dependency on `os_pipe`

**Review by rbonichon — APPROVED** [review]

File: optimism/src/preimage_oracle.rs (hunk 2 above)
Comment (mrmr1993) [reply optimism/src/preimage_oracle.rs:77]: I'm sure somebody will try at some point :)

Comment (rbonichon) [reply optimism/src/preimage_oracle.rs:77]: Right. You could craft a carefully worded comment for those uses as for the Mac users 😈

**Review by dannywillems — APPROVED** [review]
Approving, but still testing locally to see if it works.

File: optimism/src/preimage_oracle.rs [hunk 3]
```diff
@@ -24,6 +24,57 @@ pub struct ReadWrite<R, W> {
...
 
+// Here, we implement `os_pipe::pipe` in a way that allows us to pass flags. In particular, we
+// don't pass the `CLOEXEC` flag, because we want these pipes to survive an exec, and we set
+// `DIRECT` to handle writes as single atomic operations (up to splitting at the buffer size).
+// This fixes the IPC hangs. This is bad, but the hang is worse.
+
+#[cfg(not(any(target_os = "ios", target_os = "macos", target_os = "haiku", windows)))]
+fn create_pipe() -> std::io::Result<(PipeReader, PipeWriter)> {
+    let mut fds: [libc::c_int; 2] = [0; 2];
+    let res = unsafe { libc::pipe2(fds.as_mut_ptr(), libc::O_DIRECT) };
```
Comment (dannywillems) [inline optimism/src/preimage_oracle.rs:35]: I cannot find what are the default flags for `os_pipe::pipe`. If you have anything, I'm interested in.

Comment (mrmr1993) [reply optimism/src/preimage_oracle.rs:35]: https://docs.rs/crate/os_pipe/latest/source/src/unix.rs

**Comment by dannywillems** [comment]
LGTM for the test fix, tested locally.

**Comment by dannywillems** [comment]
Taking care of fixing clippy

**Comment by dannywillems** [comment]
> Taking care of fixing clippy

See 02f61d27366064908675eb76df8a6998ae1ea1ce and f5af8ff1a12f1eedbbeb2d79076d2d56adc6fed5. Waiting for CI to be happy and I'll merge.

**Pull request merged**